internal/kafka/internal/services/kafka.go

```diff
@@ -147,20 +151,95 @@ func (k *kafkaService) HasAvailableCapacityInRegion(kafkaRequest *dbapi.KafkaReq
+
 	dbConn := k.connectionFactory.New()
 
 	var count int64
+
+	var kafkas []*dbapi.KafkaRequest
+
 	if err := dbConn.Model(&dbapi.KafkaRequest{}).
 		Where("region = ?", kafkaRequest.Region).
 		Where("cloud_provider = ?", kafkaRequest.CloudProvider).
 		Where("instance_type = ?", kafkaRequest.InstanceType).
-		Count(&count).Error; err != nil {
-		return false, errors.NewWithCause(errors.ErrorGeneral, err, "failed to count kafka request")
+		Scan(&kafkas).Error; err != nil {
+		return false, errors.NewWithCause(errors.ErrorGeneral, err, errMessage)
+	}
+
+	for _, kafka := range kafkas {
+		kafkaInstanceSize, e := k.kafkaConfig.GetKafkaInstanceSize(kafka.InstanceType, kafka.SizeId)
+		if e != nil {
+			return false, errors.NewWithCause(errors.ErrorInstancePlanNotSupported, e, errMessage)
+		}
+		count += int64(kafkaInstanceSize.CapacityConsumed)
+	}
+
+	kafkaInstanceSize, e := k.kafkaConfig.GetKafkaInstanceSize(kafkaRequest.InstanceType, kafkaRequest.SizeId)
+	if e != nil {
+		return false, errors.NewWithCause(errors.ErrorInstancePlanNotSupported, e, errMessage)
 	}
-	return instTypeRegCapacity == nil || count < int64(*instTypeRegCapacity), nil
+
+	count += int64(kafkaInstanceSize.CapacityConsumed)
+
+	return instTypeRegCapacity == nil || count <= int64(*instTypeRegCapacity), nil
 }
 
-func (k *kafkaService) DetectInstanceType(kafkaRequest *dbapi.KafkaRequest) (types.KafkaInstanceType, *errors.ServiceError) {
+func (k *kafkaService) GetAvailableSizesInRegion(criteria *FindClusterCriteria) ([]string, *errors.ServiceError) {
+	if criteria == nil {
+		err := errors.GeneralError("unable to get available sizes in region: criteria was not specified")
+		logger.Logger.Error(err)
+		return nil, err
+	}
+
+	supportedInstanceTypes := k.kafkaConfig.SupportedInstanceTypes.Configuration
+	instanceType, err := supportedInstanceTypes.GetKafkaInstanceTypeByID(criteria.SupportedInstanceType)
+	if err != nil {
+		err := errors.InstanceTypeNotSupported("unable to get available sizes in region: %s", err.Error())
+		logger.Logger.Error(err)
+		return nil, err
+	}
+
+	// The kafka size list configuration must always be ordered starting with the smallest unit.
+	// The following finds the largest Kafka size that is still available in this region. Anything smaller than this
+	// size will also be considered as available to create with the remaining capacity.
+	for i := len(instanceType.Sizes) - 1; i >= 0; i-- {
+		kafka := &dbapi.KafkaRequest{
+			CloudProvider: criteria.Provider,
+			Region:        criteria.Region,
+			InstanceType:  criteria.SupportedInstanceType,
+			MultiAZ:       criteria.MultiAZ,
+			SizeId:        instanceType.Sizes[i].Id,
+		}
+
+		// Check against region limits
+		hasCapacity, err := k.HasAvailableCapacityInRegion(kafka)
+		if err != nil {
+			logger.Logger.Error(err)
+			return nil, err
+		}
+
+		if hasCapacity {
```

Comment:
should there be a error returned if there is no capacity in the region

Reply:
Looking at the code I don't think that's the case. If there is no capacity then the result (the available sizes in the region) is a nil slice/without elements.

cc @JameelB might help clarify this but I think that was the original intention looking at the code.

Reply:
no error should be raised here as `no size can be placed` is a valid return.

Reply:
Yes that is the intention, and as what Peter mentioned a region not having capacity is a valid return. 

This is used for the cloud_provider/{id}/regions endpoint. If an instance type for a particular region doesn't have any capacity, the `available_sizes` should just show up as an empty list. i.e.

```
{
      "kind": "CloudRegion",
      "id": "us-east-1",
      "display_name": "US East, N. Virginia",
      "enabled": true,
      "supported_instance_types": [
        "standard"
      ],
      "capacity": [
        {
          "instance_type": "standard",
          "max_capacity_reached": true,
          "available_sizes": []
        }
      ]
    },
```